docs/topics/request-queues.rst

```diff
@@ -0,0 +1,160 @@
+
+    When request serialization fails, Scrapy falls back to using a memory
+    downstream queue.
+
+    Scrapy provides the following disk downstream queue classes:
+
+    .. autoclass:: MarshalFifoDiskQueue
+
+    .. autoclass:: MarshalLifoDiskQueue
+
+    .. autoclass:: PickleFifoDiskQueue
+
+    .. autoclass:: PickleLifoDiskQueue
+
+
+Writing your own disk downstream queue class
+--------------------------------------------
+
+If you want to define and use your own disk downstream queue implementation,
+it has to conform to the following interface:
+
+.. class:: MyExternalQueue
+
+   .. classmethod:: __init__(cls, crawler: scrapy.crawler.Crawler, key: str, state)
+
+      Initializes an instance of this disk downstream queue class.
+
+      *key* is the unique ID of the disk downstream queue instance. It may be
+      used, for example, to create a unique file or folder to store the queue
+      content.
+
+      *state* is a variable containing state of queue. It should be serializeable
+      and deserializeble by json module. You need to come up with proper type
+      and structure to completely restore an internal state of your queue.
+
+   .. method:: push(self, request: scrapy.http.Request) -> None
+
+      Push a request into the queue.
+
+      The helper function :func:`~scrapy.utils.reqser.request_to_dict` can be
+      used to convert the request into a dict that can then be easily
+      serialized with, for example, :func:`pickle.dumps`.
+
+      Scrapy falls back to the memory downstream queue for *request* if one of
+      the following exceptions is raised:
+
+      -     :exc:`TransientError`: indicates a temporary failure.
+
+            For example, if storing requests on a server, raise this exception
+            if you temporarily loose access to the server.
+
+        -   :exc:`SerializationError`: indicates that *request* could not be
+            serialized
+
+   .. method:: pop(self) -> Optional[Request]
+
+      Pop a request from the queue.
+
+      In case of a temporary problem or empty queue, ``None`` is returned.
+      Exceptions raised in this method woudln't be handled and may lead to
+      halt of the crawler.
+
+      The helper function :func:`~scrapy.utils.reqser.request_from_dict` can
+      be used to convert a deserialized dict back into a
+      :class:`~scrapy.http.Request` object.
+
+   .. method:: peek(self) -> Optional[Request]
+
+      Returns the next object to be returned by :meth:`pop`,
+      but without removing it from the queue.
+
+      Raises :exc:`NotImplementedError` if the underlying queue class does
+      not implement a ``peek`` method, which is optional for queues.
+
+      In case of a temporary problem or empty queue, ``None`` is returned.
+
+   .. method:: close(self) -> Optional[Any]
+
+      Release internal resources (e.g. close files or sockets) and return
+      current internal state if any. This state will be seriliazed and passed
+      to __init__ method.
```

Comment:
Here we can include the detail about the data needing to be JSON-serializable, as well as a reference to the topic about pausing and resuming crawls, for better context.

```suggestion
      Release internal resources (e.g. close files or sockets) and return an 
      object serializable by :function:`json.dump` (e.g. a :class:`dict`) that 
      represents the current internal state of this queue. When :ref:`resuming 
      a crawl <topics-jobs>`, this object is passed to the :method:`__init__` 
      method of this queue class to restore the queue state.
```